akka-actor/src/main/scala/akka/util/FrequencySketch.scala

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.util
+
+import scala.collection.immutable
+
+/**
+ * A frequency sketch for estimating the popularity of items. For implementing the TinyLFU cache admission policy.
+ * The frequency sketch includes the TinyLFU reset operation, which periodically halves all counters.
+ */
+private[akka] object FrequencySketch {
+
+  /**
+   * Create a new FrequencySketch based on the cache capacity (which will be increased to the nearest power of two).
+   *
+   * @param capacity the cache capacity (maximum items that will be cached)
+   * @param widthMultiplier a multiplier for the width of the sketch
+   * @param resetMultiplier the multiplier on the capacity for the reset size
+   * @param depth the depth of count-min sketch (number of hash functions)
+   * @param counterBits the size of the counters in bits: 2, 4, 8, 16, 32, or 64 bits
+   * @param hasher the hasher builder for hash functions
+   * @return a configured FrequencySketch
+   */
+  def apply[A](
+      capacity: Int,
+      widthMultiplier: Int = 4,
+      resetMultiplier: Double = 10,
+      depth: Int = 4,
+      counterBits: Int = 4)(implicit hasher: Hasher.Builder[A]): FrequencySketch[A] = {
+    val width = widthMultiplier * Bits.ceilingPowerOfTwo(capacity)
+    val resetSize = (resetMultiplier * capacity).toInt
+    new FrequencySketch(depth, width, counterBits, resetSize, hasher, Hasher.DefaultInitialSeed)
+  }
+
+  sealed trait Hasher[A] {
+    def hash(value: A): Int
+  }
+
+  // Use a similar approach for hashing as the count-min sketch in Algebird,
+  // using MurmurHash3 with different random seeds for the pairwise independent hash functions.
+  object Hasher {
+    final val DefaultInitialSeed = 135283237
+
+    sealed trait Builder[A] {
+      def hasher(seed: Int): Hasher[A]
+    }
+
+    def generate[A](builder: Builder[A], hashers: Int, initialSeed: Int): immutable.IndexedSeq[Hasher[A]] = {
+      val random = new scala.util.Random(initialSeed)
+      (1 to hashers).map(_ => builder.hasher(seed = random.nextInt()))
+    }
+
+    implicit object StringHasher extends Hasher.Builder[String] {
+      override def hasher(seed: Int): Hasher[String] = new StringHasher(seed)
+    }
+
+    final class StringHasher(seed: Int) extends Hasher[String] {
+      override def hash(value: String): Int = scala.util.hashing.MurmurHash3.stringHash(value, seed)
```

Comment:
Didn't notice that implementation of count-min sketch. Will take a look.

Reply:
I've switched over to the key part of the approach in that implementation (to combine hash functions with the row number, rather than separate hash functions with their own seeds). Faster with it doing less hashing. According to the simple benchmark it improves throughput by 25%. The simple accuracy test also looks good.